src/kafka/connection/connection_manager.cc

```diff
@@ -44,9 +47,9 @@ std::optional<lw_shared_ptr<kafka_connection>> connection_manager::get_connectio
 future<> connection_manager::disconnect(const connection_id& connection) {
     auto conn = _connections.find(connection);
     if (conn != _connections.end()) {
-        return conn->second->close().then([this, conn] {
-            _connections.erase(conn);
-        });
+        auto conn_ptr = conn->second;
+        _connections.erase(conn);
+        return conn_ptr->close().finally([conn_ptr]{});
```

Comment:
- `.finally([conn_ptr]{});`
`conn_ptr` is `lw_shared_ptr`, so after the function returns, it goes out-of-scope (without `finally`).
- `erase` before `close`
By erasing before `close`, next `connect` will be forced to establish a new connection, instead of using "old" connection, which is now closing. (`disconnect`, `close` starts, `connect` to the same host, `close` ends)